DM-2653 Add details on how the LWM2M agent sends data to decoders (#1635)
SoftwareAG/c8y-docs: content/protocol-integration/lwm2m-bundle/device-protocols.md

Turn on **Custom Actions** to map LWM2M data into {{< product-c8y-iot >}} using custom data processing actions. For specialized integration use cases, it is required to perform customized data processing on LWM2M data. One example are LWM2M resources of the OPAQUE data type that contain proprietary, binary data, CBOR, XML or alike.

> We here distinguish between predefined custom actions and decoder microservices

Reviewer: ```suggestion
We distinguish here between predefined custom actions and decoder microservices.
```

![Custom action example](/images/device-protocols/lwm2m/lwm2m-custom-action-connectivity-signal.png)

##### Decoder microservices

> {{< product-c8y-iot >}} LWM2M allows the set of custom actions to be extended using decoder microservices. A decoder microservice is an ordinary {{< product-c8y-iot >}} microservice that implements a simple decoder interface. The LWM2M agent calls this microservice for decoding data in a customer-specific way. We are providing an according example how to write such a decoder microservice in our public [GitHub repository](https://github.com/SoftwareAG/cumulocity-examples).

Reviewer: ```suggestion
{{< product-c8y-iot >}} LWM2M allows the set of custom actions to be extended using decoder microservices. A decoder microservice is an ordinary {{< product-c8y-iot >}} microservice that implements a simple decoder interface. The LWM2M agent calls this microservice for decoding data in a customer-specific way. We provide an according example how to write such a decoder microservice in our public [GitHub repository](https://github.com/SoftwareAG/cumulocity-examples).
```

The LWM2M agent serializes different LWM2M data types to binary data depending on the type of the corresponding resource:

<table>
<tr>
<th>LWM2M resource type</th>
<th>Length (bytes)</th>
<th>Byte Order</th>
<th width="30%">Comment</th>
</tr>

<tr>
<td>OPAQUE</td>
<td>dynamic</td>
<td>unchanged</td>
<td>OPAQUE data is used to transmit raw binary blobs between a LWM2M device and a LWM2M server. The LWM2M agent simply forwards this data to the decoder microservice. The length and the order of the bytes remain unchanged.</td>
</tr>

<tr>
<td>INTEGER</td>
<td>8</td>
<td>Big Endian</td>
<td></td>
</tr>

<tr>
<td>FLOAT</td>
<td>8</td>
<td>Big Endian</td>
<td></td>
</tr>

<tr>
<td>STRING</td>
<td>dynamic</td>
<td></td>

> <td>The string is serialized into a series of bytes, starting from the first character in the string. The first byte corresponds to its first character. Likewise, the last byte corresponds to the last character of the string </td>

Reviewer: ```suggestion
<td>The string is serialized into a series of bytes, starting from the first character in the string. The first byte corresponds to its first character. Likewise, the last byte corresponds to the last character of the string. </td>
```
